tss-esapi/src/context/tpm_commands/duplication_commands.rs

```diff
@@ -1,9 +1,334 @@
 // Copyright 2021 Contributors to the Parsec project.
 // SPDX-License-Identifier: Apache-2.0
 use crate::Context;
+use crate::{
+    handles::ObjectHandle,
+    structures::{Data, EncryptedSecret, Private, SymmetricDefinitionObject},
+    tss2_esys::*,
+    Error, Result,
+};
+use log::error;
+
+use std::convert::TryFrom;
+use std::ptr::null_mut;
 
 impl Context {
-    // Missing function: Duplicate
+    /// Duplicate a loaded object so that it may be used in a different hierarchy.
+    ///
+    /// # Details
+    /// This command duplicates a loaded object so that it may be used in a different hierarchy.
+    /// The new parent key for the duplicate may be on the same or different TPM or the Null hierarchy.
+    /// Only the public area of `new_parent_handle` is required to be loaded.
+    ///
+    /// # Arguments
+    /// * `object_to_duplicate` - An [ObjectHandle] of the object that will be duplicated.
+    /// * `new_parent_handle` - An [ObjectHandle] of the new parent.
+    /// * `encryption_key_in` - An optional encryption key.
+    /// * `symmetric_alg` - Symmetric algorithm to be used for the inner wrapper.
+    ///
+    /// The `object_to_duplicate` need to be have Fixed TPM and Fixed Parent attributes set to `false`.
+    ///
+    /// # Returns
+    /// The command returns a tuple consisting of:
+    /// * `encryption_key_out` - TPM generated, symmetric encryption key for the inner wrapper if
+    ///   `symmetric_alg` is not `Null`.
+    /// * `duplicate` - Private area that may be encrypted.
+    /// * `out_sym_seed` - Seed protected by the asymmetric algorithms of new parent.
+    ///
+    /// ```rust
+    /// use std::convert::{TryFrom, TryInto};
```

Comment:
Really nice example!
It is nice that it runs and works as test. Though this will show up as an example in the documentation and there it will probably be way to long. So what we have done previously is to hide everything that does not directly involve calling the api that you are documenting. So that we only show the important parts but still the documentation will fail to compile if we change something (if you need some inspiration look at evict_control, that have like 300 lines of code but only about 20 is visible).

Reply:
Understood. I was thinking about reducing the visible lines but there are quite some things that need to be set (like fixedTPM and fixedParent = false) for the example to work... I'll trim it down.